File: demo/rust/src/agent_demo.rs
// mq9 Rust SDK — Agent Demo
//
// Demonstrates:
//   1. Agent registers its capabilities
//   2. Agent sends heartbeat via report
//   3. Discover by full-text search
//   4. Discover by semantic search
//   5. Send a task to discovered agent's mailbox
//   6. Agent unregisters at shutdown

use mq9::{Mq9Client, SendOptions};
use serde_json::json;

const SERVER: &str = "nats://demo.robustmq.com:4222";

#[tokio::main]
async fn main() -> mq9::Result<()> {
    let client = Mq9Client::connect(SERVER).await?;

    // ── 1. Create mailbox for the agent ─────────────────────────────────
    let address = client.mailbox_create(Some("demo.rust.translator"), 300).await?;
    println!("[mailbox] agent mailbox: {}", address);

    // ── 2. Register agent ────────────────────────────────────────────────
    client.agent_register(json!({
        "name": "demo.rust.translator",
        "mailbox": address,
        "payload": "Multilingual translation agent. Supports EN, ZH, JA, KO. Input: text + target language. Output: translated text."
    })).await?;
    println!("[register] agent registered: demo.rust.translator");

    // ── 3. Send heartbeat ────────────────────────────────────────────────
    client.agent_report(json!({
        "name": "demo.rust.translator",
        "mailbox": address,
        "report_info": "running, processed: 512 tasks, avg latency: 210ms"
    })).await?;
    println!("[report] heartbeat sent");

    // ── 4. Discover by full-text search ──────────────────────────────────
    let by_text = client.agent_discover(Some("translator"), None, Some(5), Some(1)).await?;
    println!("\n[discover] text='translator' → {} result(s):", by_text.len());
    for a in &by_text {
        println!("  name={}  mailbox={}", a["name"], a["mailbox"]);
    }

    // ── 5. Discover by semantic search ───────────────────────────────────
    let by_semantic = client.agent_discover(
        None,
        Some("I need to translate Chinese text into English"),
        Some(5),
        Some(1),
    ).await?;
    println!("\n[discover] semantic='translate Chinese to English' → {} result(s):", by_semantic.len());
    for a in &by_semantic {
        println!("  name={}  mailbox={}", a["name"], a["mailbox"]);
    }

    // ── 6. Send a task to discovered agent ───────────────────────────────
    if let Some(first) = by_semantic.first() {
        if let Some(target) = first["mailbox"].as_str() {
            let reply_address = client.mailbox_create(None, 60).await?;
            let payload = json!({
                "text": "你好，世界",
                "target_lang": "en",
                "reply_to": reply_address
            });
            let msg_id = client.send(target, payload.to_string(), SendOptions::default()).await?;
            println!("\n[send] task sent to {}  msg_id={}", target, msg_id);
            println!("[send] reply_to={}", reply_address);
        }
    }

    // ── 7. List all agents ────────────────────────────────────────────────
    let all = client.agent_discover(None, None, Some(20), Some(1)).await?;
    println!("\n[discover] all agents → {} registered", all.len());

    // ── 8. Unregister ─────────────────────────────────────────────────────
    client.agent_unregister(&address).await?;
    println!("\n[unregister] agent {} unregistered", address);

    client.close().await?;
    Ok(())
}

File: demo/go/agent_demo.go
// mq9 Go SDK — Agent Demo
//
// Demonstrates:
//  1. Agent registers its capabilities
//  2. Agent sends heartbeat via report
//  3. Discover by full-text search
//  4. Discover by semantic search
//  5. Send a task to discovered agent's mailbox
//  6. Agent unregisters at shutdown
//
// Run: go run agent_demo.go

package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"

	mq9 "github.com/robustmq/mq9/go"
)

const server = "nats://demo.robustmq.com:4222"

func main() {
	client, err := mq9.Connect(server)
	if err != nil {
		log.Fatal(err)
	}
	defer client.Close()

	ctx := context.Background()

	// ── 1. Create mailbox for the agent ─────────────────────────────────
	address, err := client.MailboxCreate(ctx, "demo.go.translator", 300)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("[mailbox] agent mailbox: %s\n", address)

	// ── 2. Register agent ────────────────────────────────────────────────
	err = client.AgentRegister(ctx, map[string]any{
		"name":    "demo.go.translator",
		"mailbox": address,
		"payload": "Multilingual translation agent. Supports EN, ZH, JA, KO. Input: text + target language. Output: translated text.",
	})
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println("[register] agent registered: demo.go.translator")

	// ── 3. Send heartbeat ────────────────────────────────────────────────
	client.AgentReport(ctx, map[string]any{
		"name":        "demo.go.translator",
		"mailbox":     address,
		"report_info": "running, processed: 256 tasks, avg latency: 280ms",
	})
	fmt.Println("[report] heartbeat sent")

	// ── 4. Discover by full-text search ──────────────────────────────────
	byText, _ := client.AgentDiscover(ctx, "translator", "", 5, 1)
	fmt.Printf("\n[discover] text='translator' → %d result(s):\n", len(byText))
	for _, a := range byText {
		fmt.Printf("  name=%v  mailbox=%v\n", a["name"], a["mailbox"])
	}

	// ── 5. Discover by semantic search ───────────────────────────────────
	bySemantic, _ := client.AgentDiscover(ctx, "", "I need to translate Chinese text into English", 5, 1)
	fmt.Printf("\n[discover] semantic='translate Chinese to English' → %d result(s):\n", len(bySemantic))
	for _, a := range bySemantic {
		fmt.Printf("  name=%v  mailbox=%v\n", a["name"], a["mailbox"])
	}

	// ── 6. Send a task to discovered agent ───────────────────────────────
	if len(bySemantic) > 0 {
		target, _ := bySemantic[0]["mailbox"].(string)
		if target != "" {
			replyAddress, _ := client.MailboxCreate(ctx, "", 60)
			payload, _ := json.Marshal(map[string]any{
				"text":        "你好，世界",
				"target_lang": "en",
				"reply_to":    replyAddress,
			})
			msgID, _ := client.Send(ctx, target, payload, mq9.SendOptions{})
			fmt.Printf("\n[send] task sent to %s  msg_id=%d\n", target, msgID)
			fmt.Printf("[send] reply_to=%s\n", replyAddress)
		}
	}

	// ── 7. List all agents ────────────────────────────────────────────────
	all, _ := client.AgentDiscover(ctx, "", "", 20, 1)
	fmt.Printf("\n[discover] all agents → %d registered\n", len(all))

	// ── 8. Unregister ─────────────────────────────────────────────────────
	client.AgentUnregister(ctx, address)
	fmt.Printf("\n[unregister] agent %s unregistered\n", address)
}
